### ISSUE TITLE
support for AngularJS 2

### ISSUE THREAD

**Issue by nccchirag** [RadMie/ng-knob#9]
Does this directive support AngularJS 2 beta?

**Comment by thibautLedz** [comment]
I'm on it: https://github.com/RadMie/ng-knob/pull/23

**Comment by hounsellal** [comment]
@thiboot : I can't figure out how to install/use your angular2 version in Ionic2. I have:
1) downloaded the d3 folder (which has d3.min etc.)
2) downloaded your ng-knob2.ts script.
3) imported Knob from that script in my page's TS file.

It doesn't work. The app won't build the app.js file. Can you help?

**Comment by thibautLedz** [comment]
@hounsellal :

To import d3 with angular2 you needs the d3 typings:
`npm install -g typings`
`typings search d3`
`typings install d3 --save`
- http://blog.ionic.io/ionic-and-typings/
- https://forum.ionicframework.com/t/d3js-import-typescript/56027/2

After that, you just need to:
1) Download ng-knob2.ts in your app folder
2) Import it in your component and declare it as a directive
example:

```
import {Component, OnInit, AfterViewInit} from '@angular/core';
import {NavController, NavParams} from 'ionic-angular';

import {AppSettings} from '../../appSettings';
import {Knob} from './share/ng-knob2';

@Component({
  selector: 'meteo-sensors',
  templateUrl: 'build/sensors/meteo-sensors/meteo-sensors.component.html',
  directives: [Knob]
})
export class MeteoSensorsComponent implements OnInit, AfterViewInit {
  private _DHT22KnobHumidityOptions: {};
  private _DHT22KnobHumidityValue: number = null;

  constructor(private _http: Http) {
    this._DHT22KnobHumidityOptions = {
      skin: {
        type: 'tron',
        width: 4,
        color: AppSettings.COLOR_PRIMARY,
        spaceWidth: 2
      },
      readOnly: true,
      size: 200,
      animate: {
        enabled: true,
        duration: 2000,
        ease: 'exp'
      },
      unit: "%",
      barColor: AppSettings.COLOR_PRIMARY,
      trackWidth: 30,
      barWidth: 30,
      textColor: AppSettings.COLOR_LIGHT,
      prevBarColor: 'rgba(0,156,255,.2)',
      trackColor: 'rgba(0,156,255,.2)',
      fontSize: 'auto',
      subText: {
        enabled: true,
        text: "Humidity",
        color: AppSettings.COLOR_PRIMARY,
        font: "auto"
      },
      step: 1,
      displayPrevious: true,
      min: 0,
      max: 100,
      notifyOnDragEnd: true
    };
  }
```

In your template html:
`<ui-knob *ngIf="_DHT22KnobHumidityValue" [options]="_DHT22KnobHumidityOptions" [(value)]="_DHT22KnobHumidityValue"></ui-knob>`

**Comment by hounsellal** [comment]
@thiboot :
Thanks very much. The typings installation on its own didn't work. I finally got it to work, but this is what I had to do:
npm install d3@"<4" --save 
For that step, "npm install d3" was installing v4 which didn't work properly with the ng-knob.
Everything's working now though with d3 3.5.7.

But I wasn't declaring it as a directive, so that's another problem you pointed out. Thanks very much.

**Comment by hounsellal** [comment]
But I wasn't declaring it as a directive, so that's another problem you pointed out. Thanks very much.

**Comment by pep2** [comment]
Hello,

Following the instructions above. Where do I download ng-knob2.ts from?

And how do I install it if not using bower?

(Am using Angular2)

**Comment by nccchirag** [comment]
https://github.com/thiboot/ng-knob/tree/master/src

Best,
@nccchirag https://twitter.com/nccchirag

**Comment by HJPepper** [comment]
thank you :-)

**Comment by pep2** [comment]
Thank you your response. Sorry to come back to you, but have had a Google and am stuck. Have done all the above. My app runs, but the dial is not displaying, and am getting an error. 

> Argument of type '{ selector: string; template: any; styles: any[]; directives: typeof Knob[]; }' is not assignable to parameter of type 'Component'.
>   Object literal may only specify known properties, and 'directives' does not exist in type 'Component'.

**Comment by hounsellal** [comment]
Any idea how to get this working on ionic2 rc3 which no longer uses typings?

**Comment by hounsellal** [comment]
For other looking for answers here's what I did in Ionic2 RC3:
npm install d3@"<4" --save
changed the ng-knob2.ts d3 import line to be:
import d3 from 'd3'
In app.module.ts:
import {Knob} from '../share/ng-knob2' (depending on where you saved the ng-knob2 file)
add Knob to declarations and to schemas also in that file
install typings by:
npm install @types/d3@'^3.5.17' --save-dev --save-exact

I think that's what I did and it now works.

**Comment by olivermuc** [comment]
Hi everyone & @RadMie  

I just submitted via email, in lack of a proper GIT request, a combined version of the D3 v4 support fixes + angular2 version @thiboot - works like a charm (easy setup & no errors).

Let's keep our fingers crossed that @RadMie finds the time to accept/work in the changes for an official D3v4 + angular2 version.

**Comment by distante** [comment]
@olivermuc do you have a fork?

**Comment by olivermuc** [comment]
@RadMie 
no, but I did send you the source code via email - hope that works, too.

**Comment by markterrill** [comment]
Any forks available? or could you dropbox a zip? Thanks in advance!

**Comment by olivermuc** [comment]
Currently traveling @markterrill - but will can provide dropbox link upon return. 

Unfortunately it'll be no sooner than April 4th. 

@RadMie has the code (per my last mail), perhaps he can make it available sooner.

**Comment by markterrill** [comment]
Thanks!

**Comment by markterrill** [comment]
@olivermuc hi, wondering if I could grab that code if possible

**Comment by markterrill** [comment]
i've taken a good swing at integrating the d3 (v4) suggestions from https://github.com/RadMie/ng-knob/issues/30 and @thiboot 's work on angular2, but keep getting typescript errors  as the 'attr' function expects a string.

` svg.selectAll("line")
                    .data(data)
                    .enter().append("line")
                    .attr({
                        x1: function (d: { x1 }) {
                            return d.x1;
                        },
                        y1: function (d: { y1 }) {
                            return d.y1;
                        },
                        x2: function (d: { x2 }) {
                            return d.x2;
                        },
                        y2: function (d: { y2 }) {
                            return d.y2;
                        },
                        "stroke-width": this.options.scale.width,
                        "stroke": this.options.scale.color
                    });`

**Comment by markterrill** [comment]
it'd be appreciated if someone @olivermuc @RadMie  etc could flick me some working code as i've grown rather attached to the ng-knob!
